### PR TITLE
Fix HashTable parallel join TSAN error

### PR DESCRIPTION
**Pull request by Yuhta**
Summary: Currently hash table insertion for one partition can probe into next partition.  Although during the actual insertion we detect this case and put the row in overflows instead, TSAN sees this as an error and we are also wasting time probing space that we will never use.  This change fixes this by early return from probing if we are stepping out of the partition.

Differential Revision: D42469145

### CODE REVIEW

File: velox/exec/HashTable.cpp [hunk 1]
```diff
@@ -189,7 +168,33 @@ class ProbeState {
           }
           return group_;
         }
-        continue;
+      }
+      uint16_t empty = simd::toBitMask(tagsInTable_ == kEmptyGroup) & kFullMask;
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.cpp:172]: nit : leave an empty line at l172?

File: velox/exec/HashTable.cpp [hunk 2]
```diff
@@ -189,7 +168,33 @@ class ProbeState {
...
+      if (empty > 0) {
+        if (op == Operation::kProbe) {
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.cpp:174]: I assume there is not logic change in this code block? Thanks!

File: velox/exec/HashTable.cpp [hunk 3]
```diff
@@ -151,36 +152,14 @@ class ProbeState {
...
-          if (indexInTags_ != kNotSet) {
-            // We came to the end of the probe without a hit. We replace the
-            // first tombstone on the way.
-            --numTombstones;
-            return insert(row_, insertTagIndex + indexInTags_);
-          }
-          //--numTombstoneGroups;
-          auto pos = bits::getAndClearLastSetBit(empty);
-          return insert(row_, tagIndex_ + pos);
+      if constexpr (op == Operation::kInsert) {
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.cpp:155]: Nice refactoring and fix!

File: velox/exec/HashTable.cpp [hunk 4]
```diff
@@ -995,6 +1000,14 @@ FOLLY_ALWAYS_INLINE void HashTable<ignoreNullKeys>::buildFullProbe(
     int32_t partitionBegin,
     int32_t partitionEnd,
     std::vector<char*>* FOLLY_NULLABLE overflows) {
+  auto insert = [&](int32_t /*row*/, int32_t index) {
```
Comment (xiaoxmeng) [inline velox/exec/HashTable.cpp:1003]:
```
s/insert/insertFn or insertCB/
```

**Review by xiaoxmeng — APPROVED** [review]
@Yuhta LGTM. Thanks for the fix!

File: velox/exec/HashTable.cpp (hunk 2 above)
Comment (Yuhta) [reply velox/exec/HashTable.cpp:174]: There is no logic change here, the only logic change is line 155-161

**Pull request closed without merging**